File: Code Generation/responses/python/claude/medium/sum-of-k-subarrays-with-length-at-least-m.py
from typing import List

class Solution:
    def maxSum(self, nums: List[int], k: int, m: int) -> int:
        n = len(nums)
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]
        
        def subarray_sum(l, r):  # inclusive
            return prefix[r + 1] - prefix[l]
        
        # dp[j][i] = max sum using j subarrays from nums[0..i-1]
        # We use i as "first i elements considered"
        NEG_INF = float('-inf')
        
        # dp[j][i]: max sum picking j non-overlapping subarrays each of length >= m
        # from nums[0..i-1]
        # Transition: either we don't use nums[i-1] in any new subarray ending at i,
        # or we pick a subarray ending at index i-1 with length l >= m
        
        # dp[j][i] = max(dp[j][i-1], max over l>=m of (dp[j-1][i-l] + sum(i-l..i-1)))
        
        dp = [[NEG_INF] * (n + 1) for _ in range(k + 1)]
        dp[0][0] = 0
        for i in range(1, n + 1):
            dp[0][i] = 0
        
        for j in range(1, k + 1):
            # For each i, we need max over l in [m, i-...] of dp[j-1][i-l] + prefix[i] - prefix[i-l]
            # = prefix[i] + max over l>=m of (dp[j-1][i-l] - prefix[i-l])
            # Let t = i - l, so t ranges from 0 to i-m
            # = prefix[i] + max over t in [0, i-m] of (dp[j-1][t] - prefix[t])
            
            best = NEG_INF  # max of dp[j-1][t] - prefix[t] for t in [0, i-m]
            
            for i in range(1, n + 1):
                # Update best with t = i - m (when we've moved far enough)
                t = i - m
                if t >= 0 and dp[j-1][t] != NEG_INF:
                    best = max(best, dp[j-1][t] - prefix[t])
                
                # Option 1: don't end a subarray at position i
                dp[j][i] = dp[j][i-1]
                
                # Option 2: end a subarray at position i (1-indexed, so at nums[i-1])
                if best != NEG_INF:
                    dp[j][i] = max(dp[j][i], prefix[i] + best)
        
        return dp[k][n]

File: Code Generation/responses/cpp/claude/medium/sum-of-k-subarrays-with-length-at-least-m.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long maxSum(vector<int> nums, int k, int m) {
        int n = nums.size();
        vector<long long> prefix(n + 1, 0);
        for (int i = 0; i < n; i++)
            prefix[i + 1] = prefix[i] + nums[i];
        
        // dp[j][i] = max sum using j subarrays from first i elements
        // Use rolling array: prev and curr
        const long long NEG_INF = LLONG_MIN / 2;
        
        vector<vector<long long>> dp(k + 1, vector<long long>(n + 1, NEG_INF));
        for (int i = 0; i <= n; i++) dp[0][i] = 0;
        
        for (int j = 1; j <= k; j++) {
            // best[i] = max dp[j-1][t] for t <= i
            // We want: dp[j][i] = max over l in [m, i] of (dp[j-1][i-l] + prefix[i] - prefix[i-l])
            // = prefix[i] + max over t in [0, i-m] of (dp[j-1][t] - prefix[t])
            
            long long best = NEG_INF;
            for (int i = 0; i <= n; i++) {
                // At position i, we can end a subarray here
                // The subarray ends at index i (1-based), starts at i-l+1
                // We need i >= m*j (enough elements for j subarrays)
                
                // Update best with dp[j-1][i-m] - prefix[i-m] if valid
                if (i >= m) {
                    long long val = dp[j-1][i-m];
                    if (val != NEG_INF)
                        best = max(best, val - prefix[i-m]);
                }
                
                if (best != NEG_INF)
                    dp[j][i] = max(dp[j][i], prefix[i] + best);
            }
        }
        
        long long ans = NEG_INF;
        for (int i = 0; i <= n; i++)
            ans = max(ans, dp[k][i]);
        return ans;
    }
};